File: Difficulty: Medium/Minimize the Heights II/minimize-the-heights-ii.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

  public class Main {

    public static void main(String[] args) throws Exception {
        BufferedReader br =
            new BufferedReader(new InputStreamReader(System.in));
        int tc = Integer.parseInt(br.readLine().trim());
        while (tc-- > 0) {
            String[] inputLine;
            inputLine = br.readLine().trim().split(" ");
            int k = Integer.parseInt(inputLine[0]);
            inputLine = br.readLine().trim().split(" ");
            int n = Integer.parseInt(inputLine[0]);
            int[] arr = new int[n];
            inputLine = br.readLine().trim().split(" ");
            for (int i = 0; i < n; i++) {
                arr[i] = Integer.parseInt(inputLine[i]);
            }

            int ans = new Solution().getMinDiff(arr, n, k);
            System.out.println(ans);
        }
    }
}
// } Driver Code Ends


// User function Template for Java

class Solution {
    int getMinDiff(int[] arr, int n, int k) {
        Arrays.sort(arr);
        
        int diff = arr[n-1]-arr[0];
        int small = arr[0]+k;
        int large =  arr[n-1]-k;
        
        for(int  i =0; i<arr.length-1;i++){
            int min =  Math.min(small, arr[i+1]-k);
            int max = Math.max(large, arr[i]+k);
            if(min < 0)continue;
            
            diff  =  Math.min(diff,max-min);
            
        }
        return diff;
    }
}

File: Difficulty: Medium/Minimize the Heights II/minimize-the-heights-ii.cpp
//{ Driver Code Starts
// Initial template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function template for C++

class Solution {
public:
    int getMinDiff(vector<int>& arr, int k) {
        int n = arr.size();
        if (n == 1) return 0; // No difference if there's only one element

        sort(arr.begin(), arr.end());

        // Initial difference between max and min
        int diff = arr[n-1] - arr[0];
        
        int smallest = arr[0] + k;
        int largest = arr[n-1] - k;

        for (int i = 1; i < n; i++) {
            // If subtracting k makes the element negative, skip
            if (arr[i] - k < 0) continue;
            
            // Update the new smallest and largest
            int min_element = min(smallest, arr[i] - k);
            int max_element = max(largest, arr[i-1] + k);
            
            // Minimize the difference
            diff = min(diff, max_element - min_element);
        }

        return diff;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        int n, k;
        cin >> k;
        cin.ignore();
        vector<int> a, b, c, d;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int num;
        while (ss >> num)
            a.push_back(num);

        Solution ob;
        auto ans = ob.getMinDiff(a, k);
        cout << ans << "\n";
    }
    return 0;
}
// } Driver Code Ends